core/src/integration/java/com/radixdlt/integration/targeted/consensus/DelayAfterFormingTcAndFormAQcTest.java

```diff
@@ -62,104 +62,94 @@
  * permissions under this License.
  */
 
-package com.radixdlt.consensus.bft;
-
-import com.radixdlt.consensus.QuorumCertificate;
-import com.radixdlt.consensus.TimeoutCertificate;
-import java.util.Objects;
-
-/** The result of a round/round voting (either QC or TC). */
-public sealed interface RoundVotingResult {
-
-  static FormedQC qc(QuorumCertificate qc) {
-    return new FormedQC(qc);
-  }
-
-  static FormedTC tc(TimeoutCertificate tc) {
-    return new FormedTC(tc);
+package com.radixdlt.integration.targeted.consensus;
+
+import static com.radixdlt.harness.predicates.NodesPredicate.anyAtExactlyStateVersion;
+import static org.junit.Assert.assertEquals;
+
+import com.google.common.collect.ImmutableList;
+import com.radixdlt.consensus.EpochNodeWeightMapping;
+import com.radixdlt.consensus.Proposal;
+import com.radixdlt.consensus.bft.Round;
+import com.radixdlt.environment.deterministic.network.MessageMutator;
+import com.radixdlt.environment.deterministic.network.MessageSelector;
+import com.radixdlt.harness.deterministic.DeterministicTest;
+import com.radixdlt.harness.deterministic.PhysicalNodeConfig;
+import com.radixdlt.modules.FunctionalRadixNodeModule;
+import com.radixdlt.modules.StateComputerConfig;
+import com.radixdlt.monitoring.Metrics;
+import org.junit.Test;
+
+// spotless:off
+/**
+ * Test a scenario where a timeout cert is formed first, but its
+ * processing is delayed, then a QC is formed
+ * which results in a QC output (RoundQuorumReached event with a QC).
+ *
+ * The test works by dropping a proposal to 3 out of 4 nodes,
+ * which results in:
+ * a) 1 node receives a proposal, sends a regular vote and then a timeout vote
+ * b) 3 nodes don't receive a proposal and they send a timeout vote for a fallback vertex
+ *
+ * If the order of received votes is mixed, and it happens to be when
+ * using the "first" message selector (it's deterministic, but just a coincidence),
+ * then the node is able to create a TC (by using the timeout signatures) before
+ * it can create a QC (on a fallback vertex, voted on by 3 nodes).
+ *
+ * Example order of received votes:
+ *  1) vote for fallback vertex + timeout
+ *  2) vote for proposal vertex + timeout
+ *  3) vote for fallback vertex + timeout
+ *  4) vote for fallback vertex + timeout
+ *
+ *  Note that if the votes for fallback vertices were received first,
```

Comment:
Do you think it's also worth a test which either has a 50:50 split or drops the timeout messages from (eg) node 4 - and checks that a timeout certificate is formed after the delay? (ie checks that the timeout certificate behaviour still works correctly?)

Reply:
Good idea :+1: I've added a 50/50 split test